### PR TITLE
.NET 7 Upgrade

### LINKED ISSUE #2389: Cannot set headers with System.Text.Json under .net 7
Marten 5.10.1

Cannot set headers for documents/events using System.Text.Json (newtonsoft is fine) on .net 7 as a stack overflow is triggered. 

Not sure if this is something to fix on Marten's end or a an upstream regression.

Minimal repro that runs fine on .net 6:
```csharp
using Marten;
using Marten.Services.Json;

public record Data(int Id);

var store = DocumentStore.For(o => {
    o.Connection("*");
    o.UseDefaultSerialization(serializerType: SerializerType.SystemTextJson);
    o.Policies.ForAllDocuments(doc =>
    {
        doc.Metadata.Headers.Enabled = true;
    });
});

using var session = store.LightweightSession();
session.SetHeader("fail", true);
session.Store(new Data(0));
session.SaveChanges();
```

Produces a stack overflow
```
Stack overflow.
Repeat 16333 times:
--------------------------------
   at System.Text.Json.JsonSerializer.WriteCore[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.Text.Json.Utf8JsonWriter, System.__Canon ByRef, System.Text.Json.Serialization.Metadata.JsonTypeInfo`1<System.__Canon>)
   at System.Text.Json.JsonSerializer.Serialize[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.Text.Json.Utf8JsonWriter, System.__Canon, System.Text.Json.JsonSerializerOptions)
   at Marten.Services.Json.JsonSerializerExtensions+DynamicObjectConverter.Write(System.Text.Json.Utf8JsonWriter, System.Object, System.Text.Json.JsonSerializerOptions)
   at System.Text.Json.Serialization.JsonConverter`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].TryWrite(System.Text.Json.Utf8JsonWriter, System.__Canon ByRef, System.Text.Json.JsonSerializerOptions, System.Text.Json.WriteStack ByRef)
   at System.Text.Json.Serialization.JsonConverter`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].WriteCore(System.Text.Json.Utf8JsonWriter, System.__Canon ByRef, System.Text.Json.JsonSerializerOptions, System.Text.Json.WriteStack ByRef)
--------------------------------
   at System.Text.Json.JsonSerializer.WriteCore[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.Text.Json.Utf8JsonWriter, System.__Canon ByRef, System.Text.Json.Serialization.Metadata.JsonTypeInfo`1<System.__Canon>)
   at System.Text.Json.JsonSerializer.Serialize[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.Text.Json.Utf8JsonWriter, System.__Canon, System.Text.Json.JsonSerializerOptions)
   at Marten.Services.Json.JsonSerializerExtensions+DynamicObjectConverter.Write(System.Text.Json.Utf8JsonWriter, System.Object, System.Text.Json.JsonSerializerOptions)
   at System.Text.Json.Serialization.JsonConverter`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].TryWrite(System.Text.Json.Utf8JsonWriter, System.__Canon ByRef, System.Text.Json.JsonSerializerOptions, System.Text.Json.WriteStack ByRef)
   at System.Text.Json.Serialization.Converters.DictionaryOfTKeyTValueConverter`3[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].OnWriteResume(System.Text.Json.Utf8JsonWriter, System.__Canon, System.Text.Json.JsonSerializerOptions, System.Text.Json.WriteStack ByRef)
   at System.Text.Json.Serialization.JsonDictionaryConverter`3[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].OnTryWrite(System.Text.Json.Utf8JsonWriter, System.__Canon, System.Text.Json.JsonSerializerOptions, System.Text.Json.WriteStack ByRef)
   at System.Text.Json.Serialization.JsonConverter`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].TryWrite(System.Text.Json.Utf8JsonWriter, System.__Canon ByRef, System.Text.Json.JsonSerializerOptions, System.Text.Json.WriteStack ByRef)
   at System.Text.Json.Serialization.JsonConverter`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].WriteCore(System.Text.Json.Utf8JsonWriter, System.__Canon ByRef, System.Text.Json.JsonSerializerOptions, System.Text.Json.WriteStack ByRef)
   at System.Text.Json.Serialization.JsonConverter`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].WriteCoreAsObject(System.Text.Json.Utf8JsonWriter, System.Object, System.Text.Json.JsonSerializerOptions, System.Text.Json.WriteStack ByRef)
   at System.Text.Json.JsonSerializer.WriteCoreAsObject(System.Text.Json.Utf8JsonWriter, System.Object, System.Text.Json.Serialization.Metadata.JsonTypeInfo)
   at System.Text.Json.JsonSerializer.WriteStringAsObject(System.Object, System.Text.Json.Serialization.Metadata.JsonTypeInfo)
   at Marten.Services.SystemTextJsonSerializer.ToJson(System.Object)
   at Marten.Internal.Operations.StorageOperation`2[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Int32, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].setHeaderParameter(Npgsql.NpgsqlParameter, Marten.Internal.IMartenSession)
   at Marten.Generated.DocumentStorage.UpsertDataOperation651275301.ConfigureParameters(Npgsql.NpgsqlParameter[], ConsoleApp1.Data, Marten.Internal.IMartenSession)
   at Marten.Internal.Operations.StorageOperation`2[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Int32, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ConfigureCommand(Weasel.Postgresql.CommandBuilder, Marten.Internal.IMartenSession)
   at Marten.Util.CommandExtensions.BuildCommand(Marten.Internal.IMartenSession, System.Collections.Generic.IEnumerable`1<Marten.Linq.QueryHandlers.IQueryHandler>)
   at Marten.Internal.UpdateBatch.ApplyChanges(Marten.Internal.IMartenSession)
   at Marten.Internal.Sessions.DocumentSessionBase.ExecuteBatch(Marten.Internal.IUpdateBatch)
   at Marten.Internal.Sessions.DocumentSessionBase.SaveChanges()
   at Program.<Main>$(System.String[])

```
From what i can tell things seem to be getting stuck in an infinite loop whenever a `JsonDynamicType` is present.

```csharp
public record Data(int Id, JsonSerializerExtensions.JsonDynamicType value);

session.Store(new Data(0, new JsonSerializerExtensions.JsonDynamicBoolean(true, ...)));
session.SaveChanges();
```

To trigger the same stack overflow.

### PR DESCRIPTION
**Pull request by oskardudycz**
- [x] Bumped to .NET 7
- [x] Removed .NET 3.1, .NET 5
- [x] Bumped C# version to 11
- [x] Removed obsolete pragmas and dead code
- [x] Used await using where possible
- [x] Updated CI
- [x] Npgsql and Weasel upgrade (see https://github.com/JasperFx/marten/pull/2394)
- [x] STJ issues
- [ ] Fix CodeGen-related issues https://github.com/JasperFx/marten/issues/2404, https://github.com/JasperFx/marten/issues/2405
- [ ] Write migration guide: https://github.com/JasperFx/marten/issues/2403

Fixes #2389.

### CODE REVIEW

**Review by jeremydmiller — APPROVED** [review]
Let's get this one in ASAP. I'm approving it as is.

@oskardudycz took one for the team here!

File: src/Marten.NodaTime/Marten.NodaTime.csproj [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
     <PropertyGroup>
         <Description>NodaTime extension for Marten</Description>
-        <VersionPrefix>6.0.0-alpha.3</VersionPrefix>
+        <VersionPrefix>6.0.0-alpha.4</VersionPrefix>
```
Comment (gfoidl) [inline src/Marten.NodaTime/Marten.NodaTime.csproj:4]: It's already set in the `Directory.Build.props`, so could be removed here as that value will be taken over.

File: src/Marten.PLv8/Marten.PLv8.csproj [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
     <PropertyGroup>
         <Description>Document transforms and patching extension for Marten</Description>
-        <VersionPrefix>6.0.0-alpha.3</VersionPrefix>
+        <VersionPrefix>6.0.0-alpha.4</VersionPrefix>
```
Comment (gfoidl) [inline src/Marten.PLv8/Marten.PLv8.csproj:5]: The same.

File: src/Marten.NodaTime/Marten.NodaTime.csproj (hunk 1 above)
Comment (oskardudycz) [reply src/Marten.NodaTime/Marten.NodaTime.csproj:4]: Thanks for the suggestion. For now we decided to keep it also here to be able to release it independently.

File: src/Marten.PLv8/Marten.PLv8.csproj (hunk 2 above)
Comment (oskardudycz) [reply src/Marten.PLv8/Marten.PLv8.csproj:5]: as above: https://github.com/JasperFx/marten/pull/2394#discussion_r1034744665

**Comment by toanxyz** [comment]
Hi @oskardudycz does this System.Text.Json support this issue?
https://github.com/JasperFx/marten/issues/2192

**Pull request merged**